apiserver/facades/client/modelupgrader/findtools.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelupgrader
+
+import (
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+	"github.com/juju/version/v2"
+
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/cloudconfig/podcfg"
+	coreos "github.com/juju/juju/core/os"
+	"github.com/juju/juju/docker"
+	envtools "github.com/juju/juju/environs/tools"
+	"github.com/juju/juju/feature"
+	"github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/state"
+	coretools "github.com/juju/juju/tools"
+)
+
+var errUpToDate = errors.AlreadyExistsf("no upgrades available")
+
+func (m *ModelUpgraderAPI) decideVersion(
+	targetVersion, agentVersion version.Number, agentStream string, st State, model Model,
+) (_ version.Number, err error) {
+	logger.Debugf("deciding target version for model upgrade, %q, %q, %q", targetVersion, agentVersion, agentStream)
+	filterMajor := agentVersion.Major
+	if targetVersion != version.Zero {
+		filterMajor = targetVersion.Major
+	}
+	streamVersions, err := m.findTools(
+		st, model, filterMajor, -1, agentVersion, "", "", agentStream,
+	)
+	if err != nil {
+		return version.Zero, errors.Trace(err)
+	}
+	if targetVersion != version.Zero {
+		if targetVersion.Compare(agentVersion.ToPatch()) < 0 {
+			return version.Zero, errUpToDate
+		}
+
+		// If not completely specified already, pick a single tools version.
+		filter := coretools.Filter{Number: targetVersion}
+		packagedAgents, err := streamVersions.Match(filter)
+		if err != nil {
+			return version.Zero, errors.Wrap(err, errors.NotFoundf("no matching agent versions available"))
+		}
+		targetVersion, packagedAgents = packagedAgents.Newest()
+		logger.Debugf("target version %q is the best version, packagedAgents %s", targetVersion, packagedAgents)
+		return targetVersion, nil
+	}
+
+	// No explicitly specified version, so find the version to which we
+	// need to upgrade. We find next available stable release to upgrade
+	// to by incrementing the minor version, starting from the current
+	// agent version and doing major.minor+1.patch=0.
+
+	// Upgrading across a major release boundary requires that the version
+	// be specified with --agent-version.
+	nextVersion := agentVersion
+	nextVersion.Minor += 1
```

Comment:
Does this cover the case where the current agent version is 2.9.20 and there's a 2.9.33 available?